internal/validator/cosignvalidator/cosign_validator.go

```diff
@@ -316,74 +355,103 @@ func (cv *CosignValidator) setupOptions(
-
-	if strings.HasPrefix(trustRoot.Key, "-----BEGIN PUBLIC KEY-----") {
-		verifier, err = sigs.LoadPublicKeyRaw([]byte(trustRoot.Key), hashAlgorithm)
-	} else {
-		verifier, err = sigs.PublicKeyFromKeyRefWithHashAlgo(ctx, trustRoot.Key, hashAlgorithm)
-	}
-
-	return KeyVerifierTuple{
-		Name:     trustRoot.Name,
-		Verifier: verifier,
-	}, err
-}
-
 // gathers the requested trust roots and turns them into a
 // verifier objects, which contains the public keys
 // that are then used for the Cosign validation
 func (cv *CosignValidator) verifiers(
 	ctx context.Context,
 	keyRefs []string,
-) ([]KeyVerifierTuple, error) {
+) ([]Verifier, error) {
 	trustRootsKeys, err := auth.GetTrustRoots(keyRefs, cv.TrustRoots, true)
 	if err != nil {
 		return nil, fmt.Errorf("error getting trust roots for validator %s: %s", cv.Name, err)
 	}
 
 	// Fall back to multiple verifiers
-	verifiers := make([]KeyVerifierTuple, 0, len(keyRefs))
+	verifiers := make([]Verifier, 0, len(keyRefs))
 	for _, trustedKey := range trustRootsKeys {
-		tuple, err := LoadKeyVerifierTuple(ctx, trustedKey)
+		verifier, err := LoadVerifier(ctx, trustedKey)
 		if err != nil {
 			return nil, err
 		}
-		verifiers = append(verifiers, tuple)
+		verifiers = append(verifiers, verifier)
 	}
 	return verifiers, nil
 }
+
+func (cv *CosignValidator) getRekorConfig(ctx context.Context) (*client.Rekor, *cosign.TrustedTransparencyLogPubKeys, error) {
+	rekorClient, err := rekor.NewClient(cv.Rekor)
+	if err != nil { // Currently cannot happen as it only fails for invalid URLs, which are already caught during unmarshalling
+		return nil, nil, fmt.Errorf("unable to create Rekor client for %s: %s", cv.Rekor, err)
+	}
+
+	if cv.RekorPubkey != "" {
+		pubs := cosign.NewTrustedTransparencyLogPubKeys()
+		err = pubs.AddTransparencyLogPubKey([]byte(cv.RekorPubkey), tuf.Active)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error adding rekor public key: %s", err)
+		}
+		return rekorClient, &pubs, nil
+	}
+
+	// env var is needed, otherwise cosign tries to create files in readonly FS, which will fail
+	_ = os.Setenv("SIGSTORE_NO_CACHE", "1")
+	pubs, err := cosign.GetRekorPubs(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting rekor public keys: %s", err)
+	}
+
+	return rekorClient, pubs, nil
+}
+
+func (cv *CosignValidator) getCTLogPubKeys(ctx context.Context) (*cosign.TrustedTransparencyLogPubKeys, error) {
+	if cv.CTLogPubkey != "" {
+		new_pubs := cosign.NewTrustedTransparencyLogPubKeys()
+		err := new_pubs.AddTransparencyLogPubKey([]byte(cv.CTLogPubkey), tuf.Active)
+		if err != nil {
+			return nil, fmt.Errorf("error adding ct log public key: %s", err)
+		}
+		return &new_pubs, nil
+	}
+	_ = os.Setenv("SIGSTORE_NO_CACHE", "1")
+	return cosign.GetCTLogPubs(ctx)
+}
+
+func (cv *CosignValidator) getFulcioCerts() (*x509.CertPool, *x509.CertPool, error) {
+	if cv.FulcioCert != "" {
+		root := x509.NewCertPool()
+		// certificate was already validated during unmarshalling
```

Comment:
Thinking about it, why do we even put the bytestring into the CosignValidator in the first place? It is unserialized into a CosignValidatorYaml, so we could make the cv.Cert and cv.FulcioCert field the cert pool with the added certificate during validation already, or am I missing something?

Reply:
yea. i changed it.